hpx/hpx_init_impl.hpp

```diff
@@ -84,10 +81,88 @@ namespace hpx
         std::at_quick_exit(detail::on_exit);
 #endif
+        return detail::run_or_start(f, params.desc_cmdline, argc, argv,
+            hpx_startup::user_main_config(params.cfg),
+            std::move(params.startup), std::move(params.shutdown), params.mode,
+            params.rp_mode, true);
+    }
+
+    /// \brief Main entry point for launching the HPX runtime system.
+    ///
+    /// This is the main entry point for any HPX application. This function
+    /// (or one of its overloads below) should be called from the users `main()`
+    /// function. It will set up the HPX runtime environment and schedule the
+    /// function given by \p f as a HPX thread.
+    inline int init(util::function_nonser<int(int, char**)> const& f, int argc,
+        char** argv, init_params const& params)
+    {
+        util::function_nonser<int(hpx::program_options::variables_map&)> main_f
+            = util::bind_back(detail::init_helper, f);
+        if (argc == 0 || argv == nullptr)
+        {
+            return init(main_f, params.dummy_argc, params.dummy_argv, params);
+        }
+        return init(main_f, argc, argv, params);
+    }
 
-        return detail::run_or_start(f, desc_cmdline, argc, argv,
-            hpx_startup::user_main_config(cfg),
-            std::move(startup), std::move(shutdown), mode, true);
+    /// \brief Main entry point for launching the HPX runtime system.
+    ///
+    /// This is the main entry point for any HPX application. This function
+    /// (or one of its overloads below) should be called from the users `main()`
+    /// function. It will set up the HPX runtime environment and schedule the
+    /// function given by \p f as a HPX thread.
+    inline int init(int argc, char** argv, init_params const& params)
+    {
+        util::function_nonser<int(hpx::program_options::variables_map&)> main_f
+            = static_cast<hpx_main_type>(::hpx_main);
+        if (argc == 0 || argv == nullptr)
+        {
+            return init(main_f, params.dummy_argc, params.dummy_argv, params);
+        }
+        return init(main_f, argc, argv, params);
+    }
+
+    /// \brief Main entry point for launching the HPX runtime system.
+    ///
+    /// This is the main entry point for any HPX application. This function
+    /// (or one of its overloads below) should be called from the users `main()`
+    /// function. It will set up the HPX runtime environment and schedule the
+    /// function given by \p f as a HPX thread.
+    inline int init(std::nullptr_t f, int argc, char** argv, init_params const& params)
+    {
+        util::function_nonser<int(hpx::program_options::variables_map&)> main_f;
+        if (argc == 0 || argv == nullptr)
+        {
+            return init(main_f, params.dummy_argc, params.dummy_argv, params);
+        }
+        return init(main_f, argc, argv, params);
+    }
+
+    /* We keep the other overloads to preserve compatibility but they will be
+    removed in the future */
```

Comment:
I think we can remove this comment if these are wrapped in `#if defined(HPX_WITH_X_COMPATIBILITY)`. Also, prefer C++ style comments (`//` instead of `/* */`).

Reply:
For now I just removed the comment cause a lot of examples / tests are still using those overloads, I'll create a macro in another PR I think

Reply:
Actually done :)